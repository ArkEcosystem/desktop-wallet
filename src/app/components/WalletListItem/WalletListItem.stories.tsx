import { Table } from "app/components/Table";
import React from "react";

import { WalletListItem } from "./WalletListItem";

export default {
	title: "Wallets / Components / WalletListItem",
};

const data = [
	{
		coin: "Btc",
		coinClassName: "text-theme-warning-400 border-theme-warning-200",
		avatarId: "test",
		address: "ASuusXSW9kfWnicScSgUTjttP6T9GQ3kqT",
		walletName: "My Wallet",
		balance: "100 BTC",
		fiat: "1,000,000 USD",
		walletTypeIcons: ["Star", "Multisig", "Ledger"],
		actions: [
			{
				label: "Action 1",
				value: "1",
			},
			{
				label: "Action 2",
				value: "2",
			},
			{
				label: "Action 3",
				value: "3",
			},
		],
	},
];

const columns = [
	{
		Header: "Asset Type",
		accessor: "avatarId",
	},
	{
		Header: "Wallet Address",
		accessor: "address",
	},
	{
		Header: "Wallet Type",
	},
	{
		Header: "Balance",
		accessor: "balance",
		className: "float-right",
	},
	{
		Header: "Fiat Value",
		accessor: "fiat",
		className: "float-right",
	},
];

export const Default = () => {
	return (
		<div>
			<div>
				<Table columns={columns} data={data}>
					{(rowData: any) => <WalletListItem {...rowData} />}
				</Table>
			</div>
		</div>
	);
};

export const SingleAction = () => {
	return (
		<div>
			<div>
				<Table columns={columns} data={data}>
					{(rowData: any) => <WalletListItem {...rowData} variant="singleAction" />}
				</Table>
			</div>
		</div>
	);
};
