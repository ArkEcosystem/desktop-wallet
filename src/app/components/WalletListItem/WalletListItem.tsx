import { Address } from "app/components/Address";
import { Avatar } from "app/components/Avatar";
import { Button } from "app/components/Button";
import { Circle } from "app/components/Circle";
import { Icon } from "app/components/Icon";
import React from "react";

import { Dropdown } from "../Dropdown";

export type WalletListItemProps = {
	coinIcon: string;
	coinClass?: string;
	address?: string;
	walletName?: string;
	balance?: string;
	fiat?: string;
	walletTypeIcons?: any[] | null;
	actions?: any[];
	variant?: "singleAction";
	onAction?: any;
};

export const WalletListItem = ({
	coinIcon,
	coinClass,
	address,
	walletName,
	balance,
	fiat,
	walletTypeIcons,
	actions,
	variant,
	onAction,
}: WalletListItemProps) => {
	const getIconTypeClass = (icon: string) => {
		if (icon === "Star") return "text-theme-warning-400";
		return "text-theme-neutral-600";
	};

	const onDropdownAction = (action: any) => {
		if (typeof onAction === "function") onAction(action);
	};

	return (
		<tr className="border-b border-theme-neutral-200">
			<td className="py-6 mt-1">
				<div className="flex">
					<Circle className={coinClass} size="lg">
						<Icon name={coinIcon} width={20} height={20} />
					</Circle>
					<Avatar size="lg" address={address as string} />
				</div>
			</td>
			<td className="py-1">
				<Address walletName={walletName} address={address} maxChars={22} />
			</td>
			{walletTypeIcons && (
				<td className="py-1 text-sm font-bold">
					{walletTypeIcons.map((type: string, index: number) => {
						return (
							<div key={index} className={`inline-block mr-2 align-middle ${getIconTypeClass(type)}`}>
								<Icon name={type} width={16} height={16} />
							</div>
						);
					})}
				</td>
			)}
			<td className="font-semibold text-right">
				<div>{balance}</div>
			</td>
			<td className="text-right text-theme-neutral-400">
				<div>{fiat}</div>
			</td>
			<td>
				{actions &&
					actions.length &&
					(() => {
						if (variant === "singleAction") {
							return (
								<div className="text-right">
									<Button data-testid="button" variant="plain" onClick={onDropdownAction}>
										{actions[0].label}
									</Button>
								</div>
							);
						}

						return (
							<div className="text-theme-neutral-400 hover:text-theme-neutral-500">
								<Dropdown options={actions} onSelect={onDropdownAction} />
							</div>
						);
					})()}
			</td>
		</tr>
	);
};

WalletListItem.defaultProps = {
	walletTypeIcons: [],
	actions: [],
	address: "",
};
