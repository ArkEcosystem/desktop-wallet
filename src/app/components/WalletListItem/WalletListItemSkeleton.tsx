import { Circle } from "app/components/Circle";
import { TableCell, TableRow } from "app/components/Table";
import React from "react";
import Skeleton from "react-loading-skeleton";

export const WalletListItemSkeleton = () => (
	<TableRow data-testid="WalletListItemSkeleton">
		<TableCell variant="start" innerClassName="space-x-4">
			<div className="-space-x-2">
				<div className="flex items-center mr-2 -space-x-1">
					<Circle className="border-transparent" size="lg">
						<Skeleton circle height={44} width={44} />
					</Circle>
					<Circle className="border-transparent" size="lg">
						<Skeleton circle height={44} width={44} />
					</Circle>
				</div>
			</div>
			<Skeleton height={16} width={260} />
		</TableCell>

		<TableCell innerClassName="justify-center text-sm font-bold text-center align-middle">
			<Skeleton height={16} width={100} />
		</TableCell>

		<TableCell innerClassName="font-semibold justify-end">
			<Skeleton height={16} width={229} />
		</TableCell>

		<TableCell variant="end" innerClassName="justify-end text-theme-secondary-400">
			<Skeleton height={16} width={100} />
		</TableCell>
	</TableRow>
);
