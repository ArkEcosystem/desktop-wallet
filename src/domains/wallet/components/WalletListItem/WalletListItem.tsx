import { ReadWriteWallet } from "@arkecosystem/platform-sdk-profiles";
import Tippy from "@tippyjs/react";
import { Address } from "app/components/Address";
import { AvatarWrapper } from "app/components/Avatar";
import { Button } from "app/components/Button";
import { Circle } from "app/components/Circle";
import { Icon } from "app/components/Icon";
import { TableCell, TableRow } from "app/components/Table";
import React from "react";
import { useTranslation } from "react-i18next";

type Option = {
	label: string;
	value: string | number;
};

type Props = {
	wallet: ReadWriteWallet;
	isCondensed: boolean;
	onAction?: (action: Option, address: string) => void;
	options: Option[];
};

export const WalletListItem = ({ wallet, isCondensed, onAction, options }: Props) => {
	const { t } = useTranslation();

	// TODO: add "Business", "Bridgechain"
	const walletTypes: string[] = ["Delegate"];

	return (
		<TableRow key={wallet.id()} border>
			<TableCell variant="start" className="w-1">
				<div className="mr-4">
					<AvatarWrapper data-testid="WalletListItem__user--avatar" size="lg" noShadow>
						<img
							src={`data:image/svg+xml;utf8,${wallet.avatar()}`}
							title={wallet.alias()}
							alt={wallet.alias()}
						/>
						{wallet && wallet.alias && (
							<span className="absolute text-sm font-semibold text-theme-background">
								{wallet?.alias()?.slice(0, 2)?.toUpperCase()}
							</span>
						)}
					</AvatarWrapper>
				</div>
			</TableCell>

			<TableCell className="">
				<Address address={wallet.address()} maxChars={24} />
			</TableCell>

			<TableCell>
				<span data-testid="WalletListItem__name">{wallet.alias()}</span>
			</TableCell>

			<TableCell>
				<span data-testid="WalletListItem__type">{t("COMMON.MY_WALLET")}</span>
			</TableCell>

			{!isCondensed && (
				<TableCell className={""} innerClassName="space-x-2 text-sm font-bold justify-center">
					{wallet.hasSyncedWithNetwork() &&
						walletTypes.map(
							(type: string) =>
								// @ts-ignore
								wallet[`is${type}`]() && (
									<Tippy key={type} content={t(`COMMON.${type.toUpperCase()}`)}>
										<Circle className="border-black" noShadow>
											<Icon name={type} width={25} height={25} />
										</Circle>
									</Tippy>
								),
						)}
				</TableCell>
			)}

			<TableCell
				variant="end"
				className="border-b border-dashed border-theme-neutral-300 dark:border-theme-neutral-800"
				innerClassName="justify-end"
			>
				<Button
					data-testid={`WalletListItem__one-option-button-${wallet.id()}`}
					className="float-right"
					variant="plain"
					onClick={() => onAction?.(options[0], wallet.address())}
				>
					{options[0]?.label}
				</Button>
			</TableCell>
		</TableRow>
	);
};

WalletListItem.defaultProps = {
	isCondensed: false,
};
