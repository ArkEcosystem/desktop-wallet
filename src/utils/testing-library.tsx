import { ARK } from "@arkecosystem/platform-sdk-ark";
import { Environment } from "@arkecosystem/platform-sdk-profiles";
import { render } from "@testing-library/react";
import { EnvironmentProvider } from "app/contexts";
import { i18n } from "app/i18n";
import { httpClient } from "app/services";
import { createMemoryHistory } from "history";
import React from "react";
import { I18nextProvider } from "react-i18next";
import { Router } from "react-router-dom";
import { StubStorage } from "tests/mocks";

import envFixture from "../tests/fixtures/env/data.json";

export const env = new Environment({ coins: { ARK }, httpClient, storage: new StubStorage(envFixture) });

const WithProviders: React.FC = ({ children }: { children?: React.ReactNode }) => (
	<I18nextProvider i18n={i18n}>
		<EnvironmentProvider env={env}>{children}</EnvironmentProvider>
	</I18nextProvider>
);

const customRender = (component: React.ReactElement, options: any = {}) =>
	render(component, { wrapper: WithProviders, ...options });

const renderWithRouter = (
	component: React.ReactElement,
	{ routes = ["/"], history = createMemoryHistory({ initialEntries: routes }), withProviders = true } = {},
) => {
	const RouterWrapper = ({ children }: { children: React.ReactNode }) =>
		withProviders ? (
			<WithProviders>
				<Router history={history}>{children}</Router>
			</WithProviders>
		) : (
			<Router history={history}>{children}</Router>
		);

	return {
		...customRender(component, { wrapper: RouterWrapper }),
		history,
	};
};

export * from "@testing-library/react";

export { customRender as render, renderWithRouter };
