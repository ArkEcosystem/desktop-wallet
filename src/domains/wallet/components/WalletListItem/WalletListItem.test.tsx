import React from "react";
import { fireEvent, render } from "testing-library";

import { WalletListItem } from "./WalletListItem";

const wallet = {
	id: () => "id5sRKWckH4rE1hQ9eeMeHAepgyC3cvJtwb",
	address: () => "D5sRKWckH4rE1hQ9eeMeHAepgyC3cvJtwb",
	alias: () => "Test Wallet",
	avatar: () => "data:image/png;base64,avatarImage",
	coinId: () => "ARK",
	networkId: () => "ark.devnet",
	isDelegate: () => true,
	isResignedDelegate: () => false,
	hasSyncedWithNetwork: () => true,
};

const options = [
	{ label: "Option 1", value: "1" },
	{ label: "Option 2", value: "2" },
];

describe("Wallet List Item", () => {
	it("should render", () => {
		const { asFragment } = render(
			<table>
				<tbody>
					<WalletListItem wallet={wallet} options={options} />
				</tbody>
			</table>,
		);

		expect(asFragment()).toMatchSnapshot();
	});

	it("should render and trigger on action on click", () => {
		const onAction = jest.fn();
		const { asFragment, getByTestId } = render(
			<table>
				<tbody>
					<WalletListItem wallet={wallet} options={options} onAction={onAction} />
				</tbody>
			</table>,
		);
		fireEvent.click(getByTestId(`WalletListItem__one-option-button-${wallet.id()}`));

		expect(onAction).toHaveBeenCalledWith({ label: "Option 1", value: "1" }, "D5sRKWckH4rE1hQ9eeMeHAepgyC3cvJtwb");
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render as condensend", () => {
		const { asFragment } = render(
			<table>
				<tbody>
					<WalletListItem wallet={wallet} isCondensed options={options} />
				</tbody>
			</table>,
		);

		expect(asFragment()).toMatchSnapshot();
	});

	it("should render without tippy", () => {
		const { asFragment } = render(
			<table>
				<tbody>
					<WalletListItem wallet={{ ...wallet, isDelegate: () => false }} options={options} />
				</tbody>
			</table>,
		);

		expect(asFragment()).toMatchSnapshot();
	});
});
