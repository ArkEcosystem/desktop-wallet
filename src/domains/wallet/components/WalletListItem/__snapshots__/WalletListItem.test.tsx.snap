// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet List Item should render 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="sc-AxheI kPnUtS group"
        data-testid="TableRow"
      >
        <td
          class="w-1"
        >
          <div
            class="sc-AxgMl eObzhI"
          >
            <div
              class="mr-4"
            >
              <div
                class="Avatar__AvatarWrapper-sc-1vsy2s2-0 hIrkRK"
                data-testid="WalletListItem__user--avatar"
              >
                <img
                  alt="Test Wallet"
                  src="data:image/svg+xml;utf8,data:image/png;base64,avatarImage"
                  title="Test Wallet"
                />
                <span
                  class="absolute text-sm font-semibold text-theme-background"
                >
                  TE
                </span>
              </div>
            </div>
          </div>
        </td>
        <td
          class=""
        >
          <div
            class="sc-AxgMl jraKvu"
          >
            <div
              class="inline-block truncate align-middle"
            >
              <span
                class="sc-AxjAm StDqN text-theme-text font-semibold text-base"
                data-testid="address__wallet-address"
              >
                D5sRKWckH4…gyC3cvJtwb
              </span>
            </div>
          </div>
        </td>
        <td>
          <div
            class="sc-AxgMl jraKvu"
          >
            <span
              data-testid="WalletListItem__name"
            >
              Test Wallet
            </span>
          </div>
        </td>
        <td>
          <div
            class="sc-AxgMl jraKvu"
          >
            <span
              data-testid="WalletListItem__type"
            >
              My Wallet
            </span>
          </div>
        </td>
        <td
          class=""
        >
          <div
            class="sc-AxgMl jraKvu space-x-2 text-sm font-bold justify-center"
          >
            <div
              class="sc-AxiKw bZmbqI border-black"
            >
              <div
                class="sc-AxhCb gZVKaf"
                height="25"
                width="25"
              >
                <svg>
                  delegate.svg
                </svg>
              </div>
            </div>
          </div>
        </td>
        <td
          class="border-b border-dashed border-theme-neutral-200"
        >
          <div
            class="sc-AxgMl iwmQfA justify-end"
          >
            <button
              class="sc-AxirZ jCYjLj float-right"
              color="primary"
              data-testid="WalletListItem__one-option-button-id5sRKWckH4rE1hQ9eeMeHAepgyC3cvJtwb"
              type="button"
            >
              Option 1
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Wallet List Item should render and trigger on action on click 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="sc-AxheI kPnUtS group"
        data-testid="TableRow"
      >
        <td
          class="w-1"
        >
          <div
            class="sc-AxgMl eObzhI"
          >
            <div
              class="mr-4"
            >
              <div
                class="Avatar__AvatarWrapper-sc-1vsy2s2-0 hIrkRK"
                data-testid="WalletListItem__user--avatar"
              >
                <img
                  alt="Test Wallet"
                  src="data:image/svg+xml;utf8,data:image/png;base64,avatarImage"
                  title="Test Wallet"
                />
                <span
                  class="absolute text-sm font-semibold text-theme-background"
                >
                  TE
                </span>
              </div>
            </div>
          </div>
        </td>
        <td
          class=""
        >
          <div
            class="sc-AxgMl jraKvu"
          >
            <div
              class="inline-block truncate align-middle"
            >
              <span
                class="sc-AxjAm StDqN text-theme-text font-semibold text-base"
                data-testid="address__wallet-address"
              >
                D5sRKWckH4…gyC3cvJtwb
              </span>
            </div>
          </div>
        </td>
        <td>
          <div
            class="sc-AxgMl jraKvu"
          >
            <span
              data-testid="WalletListItem__name"
            >
              Test Wallet
            </span>
          </div>
        </td>
        <td>
          <div
            class="sc-AxgMl jraKvu"
          >
            <span
              data-testid="WalletListItem__type"
            >
              My Wallet
            </span>
          </div>
        </td>
        <td
          class=""
        >
          <div
            class="sc-AxgMl jraKvu space-x-2 text-sm font-bold justify-center"
          >
            <div
              class="sc-AxiKw bZmbqI border-black"
            >
              <div
                class="sc-AxhCb gZVKaf"
                height="25"
                width="25"
              >
                <svg>
                  delegate.svg
                </svg>
              </div>
            </div>
          </div>
        </td>
        <td
          class="border-b border-dashed border-theme-neutral-200"
        >
          <div
            class="sc-AxgMl iwmQfA justify-end"
          >
            <button
              class="sc-AxirZ jCYjLj float-right"
              color="primary"
              data-testid="WalletListItem__one-option-button-id5sRKWckH4rE1hQ9eeMeHAepgyC3cvJtwb"
              type="button"
            >
              Option 1
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Wallet List Item should render as condensend 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="sc-AxheI kPnUtS group"
        data-testid="TableRow"
      >
        <td
          class="w-1"
        >
          <div
            class="sc-AxgMl eObzhI"
          >
            <div
              class="mr-4"
            >
              <div
                class="Avatar__AvatarWrapper-sc-1vsy2s2-0 hIrkRK"
                data-testid="WalletListItem__user--avatar"
              >
                <img
                  alt="Test Wallet"
                  src="data:image/svg+xml;utf8,data:image/png;base64,avatarImage"
                  title="Test Wallet"
                />
                <span
                  class="absolute text-sm font-semibold text-theme-background"
                >
                  TE
                </span>
              </div>
            </div>
          </div>
        </td>
        <td
          class=""
        >
          <div
            class="sc-AxgMl jraKvu"
          >
            <div
              class="inline-block truncate align-middle"
            >
              <span
                class="sc-AxjAm StDqN text-theme-text font-semibold text-base"
                data-testid="address__wallet-address"
              >
                D5sRKWckH4…gyC3cvJtwb
              </span>
            </div>
          </div>
        </td>
        <td>
          <div
            class="sc-AxgMl jraKvu"
          >
            <span
              data-testid="WalletListItem__name"
            >
              Test Wallet
            </span>
          </div>
        </td>
        <td>
          <div
            class="sc-AxgMl jraKvu"
          >
            <span
              data-testid="WalletListItem__type"
            >
              My Wallet
            </span>
          </div>
        </td>
        <td
          class="border-b border-dashed border-theme-neutral-200"
        >
          <div
            class="sc-AxgMl iwmQfA justify-end"
          >
            <button
              class="sc-AxirZ jCYjLj float-right"
              color="primary"
              data-testid="WalletListItem__one-option-button-id5sRKWckH4rE1hQ9eeMeHAepgyC3cvJtwb"
              type="button"
            >
              Option 1
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Wallet List Item should render without tippy 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="sc-AxheI kPnUtS group"
        data-testid="TableRow"
      >
        <td
          class="w-1"
        >
          <div
            class="sc-AxgMl eObzhI"
          >
            <div
              class="mr-4"
            >
              <div
                class="Avatar__AvatarWrapper-sc-1vsy2s2-0 hIrkRK"
                data-testid="WalletListItem__user--avatar"
              >
                <img
                  alt="Test Wallet"
                  src="data:image/svg+xml;utf8,data:image/png;base64,avatarImage"
                  title="Test Wallet"
                />
                <span
                  class="absolute text-sm font-semibold text-theme-background"
                >
                  TE
                </span>
              </div>
            </div>
          </div>
        </td>
        <td
          class=""
        >
          <div
            class="sc-AxgMl jraKvu"
          >
            <div
              class="inline-block truncate align-middle"
            >
              <span
                class="sc-AxjAm StDqN text-theme-text font-semibold text-base"
                data-testid="address__wallet-address"
              >
                D5sRKWckH4…gyC3cvJtwb
              </span>
            </div>
          </div>
        </td>
        <td>
          <div
            class="sc-AxgMl jraKvu"
          >
            <span
              data-testid="WalletListItem__name"
            >
              Test Wallet
            </span>
          </div>
        </td>
        <td>
          <div
            class="sc-AxgMl jraKvu"
          >
            <span
              data-testid="WalletListItem__type"
            >
              My Wallet
            </span>
          </div>
        </td>
        <td
          class=""
        >
          <div
            class="sc-AxgMl jraKvu space-x-2 text-sm font-bold justify-center"
          />
        </td>
        <td
          class="border-b border-dashed border-theme-neutral-200"
        >
          <div
            class="sc-AxgMl iwmQfA justify-end"
          >
            <button
              class="sc-AxirZ jCYjLj float-right"
              color="primary"
              data-testid="WalletListItem__one-option-button-id5sRKWckH4rE1hQ9eeMeHAepgyC3cvJtwb"
              type="button"
            >
              Option 1
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
